typing/parmatch.ml

```diff
@@ -997,15 +997,16 @@ let pats_of_type ?(always=false) env ty =
   match ty'.desc with
   | Tconstr (path, _, _) ->
       begin try match (Env.find_type path env).type_kind with
-      | Type_variant cl when always || List.length cl = 1 ||
+      | Type_variant cl when always || List.length cl <= 1 ||
+        (* Only explode when all constructors are GADTs *)
```

Comment:
I know this code predates the present PR, but can you explain why we do this only when all constructors are GADTs, instead of whenever some constructor is a GADT?

Reply:
In counter-example mode, what `type_pat` checks is that the counter-example pattern is well-typed and inhabited (non-empty). A wildcard is suspicious because it might be non-empty(precisely: there may be no search path where its type is inhabited), so we explode wildcards using `pats_of_type` to check that they contain values. If there is at least one non-gadt constructor, the type-checker assumes that this constructor is valid in the current context, and non-empty (this is a simplification that makes the check incomplete), so it doesn't try to explode the wildcard. If all constructors are GADTs, then we have to explode, and (in `type_pat`) we have to backtrack.

Consider
```
# type empty = |;;
# type 'a singleton = Single of 'a;;
# (function _ -> . : empty singleton -> 'a);;
- : empty singleton -> 'a = <fun>
# (function _ -> . : empty option -> 'a);;
Error: This match case could not be refuted.
       Here is an example of a value that would reach it: _
# (function _ -> . : (empty, empty) result -> 'a);;
Error: This match case could not be refuted.
       Here is an example of a value that would reach it: _
```

In the `empty option` case, this reasoning of the type-checker is correct (the `None` case is always inhabited). The wildcard is not split and is shown as an obscure but valid counter-example.

In the `(empty, empty) result` case, the reasoning of the type-checker is an incomplete approximation: it proposes the unexploded wildcard as a counter-example, and the user would have to explode manually to reject it.

Reply:
Thanks for the explanation.